### ISSUE TITLE
Unsure how to get a valid File type for AttachUgc.

### ISSUE THREAD

**Issue by DrBeza** [Facepunch/Facepunch.Steamworks#416]
Hi there,

Thanks for the great library.  I have been stumped on how to use `AttachUgc` with a leaderboard for a day or so now and I am unable to work out from the documentation/issues history available.

I am creating a new UGC file using:

```
var res = await Editor.NewCommunityFile
    .WithTitle("some title")
    .WithMetaData("metadata string")
    .SubmitAsync();
```
My metadata will most likely be some JSON string. I am able to retrieve this UGC based on the `FileId` result:

```
var item = await SteamUGC.QueryFileAsync(res.FileId);
```

I cannot see anything on `item` that would allow me to get a `Steamworks.Data.Ugc` type.

There was a similar issue closed without any resolution - https://github.com/Facepunch/Facepunch.Steamworks/issues/360

Is there any examples of `AttachUgc` or have I missed something simple/fundemental?

Thanks

**Comment by garrynewman** [comment]
You're right - I need to add these binds. How urgently do you need it?

**Comment by DrBeza** [comment]
Not urgently I've parked this work for the moment. Thanks Garry.

**Comment by Mat-Loz** [comment]
Looking for it as well ! Any ETA ?

**Comment by ambid17** [comment]
I am looking for this as well. I will take a look at adding it :D

**Comment by ambid17** [comment]
#488 allows you to Attach Workshop UGC to the LeaderboardEntry, however, doesn't #436 need to be implemented to actually attach a file to the Entry so that replays will function correctly?

I have tested trying to attach a UGCItem to the leaderboard item, instead of a SteamCloud file, and I've experienced intermittently working results. Replays are the last big feature I am looking to add to my game, so any info would be greatly appreciated!

**Comment by kamyker** [comment]
@ambid17 have you tried using it? Creating workshop items with FileShare is old, when I was trying it and attaching to leaderboard scores it was nothing but it was buggy - couldn't make it work.

**Comment by ambid17** [comment]
@kamyker  I did end up getting it to work!

**Comment by kamyker** [comment]
> @kamyker I did end up getting it to work!
Nice, yes I'm using something similar but I was asking about FileShare :)

**Comment by ambid17** [comment]
@kamyker I didn't end up using FileShare, it didn't seem like a viable option

**Comment by hankditton** [comment]
@ambid17 Thanks for the example, it's been really helpful!

I'm running into a problem I'm wondering if you've seen. Initially it seems like I can submit my score and attach a UGC item, then others can download that UGC item from the leaderboard entry later. 

But after a bit of time, the LeaderboardEntry no longer has a valid UGC attached (it returns 0xfffffffffful for the fileId). I have no idea why this is happening, and I'm not sure if there's something odd going on like the file timing out/expiring or what. Did you run into anything like this? 

I recently switched from having the UGC item created via NewCommunityFile and into a NewGameManagedFile (to try to address the above), but now I get failures on the upload. Were there any specific settings/permissions required to make NewGameManagedFile work?

**Comment by andefoo** [comment]
@hankditton Did you ever solve that issue of attached UGC turning into invalid file id? I might be having the same issue. 

Everything was working just fine at first but now that I added a bigger number of new leaderboard entries to different leaderboards (>30), attaching ghost recordings to them, the next day leaderboard queries are returning an invalid value for the attached UGC for those new entries. 

I also created the UGC file with NewGameManagedFile. I can still find the files with Steamworks.Ugc.Query.GameManagedItems using search text but Leaderboards are not linked to them anymore. It also seems queries don't always require exact match for the search text which partially invalidates the idea of searching for the file by unique id as title instead of using attached UGC from leaderboards. I can reupload the ghost recordings and reattach them, which seemingly fixes the issue, but I have no idea for how long, and why they got invalidated last time - or if the files will get deleted at some point as they are not attached to leaderboards.

Does anyone have any idea on what could be causing this or what should be done?

**Comment by hankditton** [comment]
@andefoo no, I didn't solve this issue but I implemented a work around that's working well. There's some details in this reddit thread, https://www.reddit.com/r/gamedev/comments/pphbhu/steamworks_leaderboard_question_vanishing_ugc. The tldr is I upload the replay file, break the file id into two ints and then manually add those ints to the details array on the leaderboard entry when the score is submitted. Then i can download a replay by reconstructing the fileid and downloading by id. 

It is working great, and has the added benefit of always having a replay when a score is uploaded (vs having to add a score first then add the file later).

**Comment by andefoo** [comment]
One more question to @hankditton or anyone else that might know the answer...

Are you sure AttachUGC is supposed to work with GameManagedItems or could it be so that you must use normal Workshop items? That would not be very nice, though, since they would then be visible in the web, too, I guess...

But anyway, it feels odd that a feature like this is so broken (or at least badly documented). I would assume there are lots of other games trying to use it, too.

**Comment by Bialy123** [comment]
@garrynewman is there any chance that AttachUGC will be working in the near future?

**Comment by Matsyir** [comment]
> But the files that you upload that way stay in the cloud forever even if they are not attached to leaderboards? Is there any way for players to delete them? Do you have any system to prevent leaking uploaded files? (For example, what happens if for some reason when making a leaderboard update deleting the old entry fails?)

@andefoo You should be able to avoid the "leaking files" issue by updating the existing leaderboard file rather than deleting it and uploading a new one.

But then, if your leaderboard entry submission fails (after updating the leaderboard file), it could leave the leaderboard entry with an invalid file / outdated entry info. This isn't as big an issue as "leaking files" though, and easily fixed by re-submitting the score.

Note: I've not tested any of this yet, just going through and preparing my Steam implementations. Probably going to follow the method mentioned by hankditton for leaderboard file attachment.

**Comment by JanMalitschek** [comment]
Hey, is this still being worked on @garrynewman ?
I've just run into this issue myself today - tried using #488 but apparently it's been deleted?
Got my replay system and everything working, I just can't do anything with it.
I would super appreciate it if this was added!
